File: src/main/java/com/nugurang/service/WorkService.java
package com.nugurang.service;

import com.nugurang.dao.ProjectDao;
import com.nugurang.dao.WorkDao;
import com.nugurang.dto.WorkInputDto;
import com.nugurang.entity.WorkEntity;
import java.util.Optional;
import org.springframework.stereotype.Service;

@Service
public class WorkService {
    private final ProjectDao projectDao;
    private final WorkDao workDao;

    public WorkEntity createWork(WorkInputDto workInputDto, Long project) {
        return workDao.save(WorkEntity.builder().name(workInputDto.getName()).opened(true).order(Optional.ofNullable(workInputDto.getOrder()).orElseGet(() -> workDao.findFirstByOrderByOrderDesc().map(prevWorkEntity -> prevWorkEntity.getOrder() + 1).orElse(0))).project(projectDao.findById(project).get()).build());
    }

    public Optional<WorkEntity> getWork(Long workId) {
        return workDao.findById(workId);
    }

    public WorkEntity updateWork(WorkInputDto workInputDto, Long workId) {
        return workDao.save(workDao.findById(workId).map(workEntity -> {
            workEntity.setName(workInputDto.getName());
            workEntity.setOpened(true);
            workEntity.setOrder(Optional.ofNullable(workInputDto.getOrder()).orElseGet(() -> workDao.findFirstByOrderByOrderDesc().map(prevWorkEntity -> prevWorkEntity.getOrder() + 1).orElse(0)));
            return workEntity;
        }).get());
    }

    public void deleteWork(Long workId) {
        workDao.deleteById(workId);
    }

    //<editor-fold defaultstate="collapsed" desc="delombok">
    @SuppressWarnings("all")
    
    public WorkService(final ProjectDao projectDao, final WorkDao workDao) {
        this.projectDao = projectDao;
        this.workDao = workDao;
    }
    //</editor-fold>
}

File: src/main/java/com/nugurang/service/WorkService.kt
package com.nugurang.service

import com.nugurang.dao.ProjectDao
import com.nugurang.dao.WorkDao
import com.nugurang.dto.WorkInputDto
import com.nugurang.entity.ProjectEntity
import com.nugurang.entity.WorkEntity
import com.nugurang.exception.NotFoundException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service

@Service
class WorkService(private val projectDao: ProjectDao, private val workDao: WorkDao) {

    fun createWork(workInputDto: WorkInputDto, projectId: Long): WorkEntity {
        return workDao.save(
            WorkEntity
            .builder()
            .name(workInputDto.name)
            .opened(true)
            .order(
                workInputDto.order ?: (
                    workDao.findFirstByOrderByOrderDesc()
                    ?.let { prevWorkEntity -> prevWorkEntity.order + 1 }
                    ?: 0
                )
            ).project(projectDao.findByIdOrNull(projectId) ?: throw NotFoundException(ProjectEntity::class.java))
            .build()
        )
    }

    fun getWork(workId: Long): WorkEntity {
        return workDao.findByIdOrNull(workId) ?: throw NotFoundException(WorkEntity::class.java)
    }

    fun updateWork(workInputDto: WorkInputDto, workId: Long): WorkEntity {
        return workDao.save(
            workDao.findByIdOrNull(workId)?.let { workEntity ->
                workEntity.name = workInputDto.name
                workEntity.opened = true
                workEntity.order = workInputDto.order ?: (
                    workDao.findFirstByOrderByOrderDesc()
                    ?.let { prevWorkEntity -> prevWorkEntity.order + 1 }
                    ?: 0
                )
                workEntity
            } ?: throw NotFoundException(WorkEntity::class.java)
        )
    }

    fun deleteWork(workId: Long) {
        workDao.deleteById(workId)
    }
}